ddtrace/contrib/aws_lambda/patch.py

```diff
@@ -121,19 +121,27 @@ def _get_handler_and_module():
     path = os.environ.get("DD_LAMBDA_HANDLER", None)
     _datadog_instrumentation = DatadogInstrumentation()
 
+    handler, handler_module = None, None
+
     if path is None:
         from datadog_lambda.wrapper import datadog_lambda_wrapper
 
+        handler_module = datadog_lambda_wrapper
         handler = getattr(datadog_lambda_wrapper, "__call__")
-
-        return handler, datadog_lambda_wrapper, _datadog_instrumentation
     else:
         parts = path.rsplit(".", 1)
         (mod_name, handler_name) = parts
         modified_mod_name = _modify_module_name(mod_name)
         handler_module = import_module(modified_mod_name)
         handler = getattr(handler_module, handler_name)
-        return handler, handler_module, _datadog_instrumentation
+
+        # Check if handler is a class instance that can be called.
+        if not hasattr(handler, "__code__") and hasattr(handler, "__call__"):
+            class_name = type(handler).__name__
+            handler_module = getattr(handler_module, class_name)
```

Comment:
`getattr(module, class_name)` this will return a reference to a class. Is this expected? If so we should rename `handler_module` or `class_name`

Reply:
Yes, this is expected. Since we don't know where this class might be coming from, we want to get the direct reference to it. We might encounter something like the AWS Chalice example, where the class `Chalice` is imported from another module. We don't have the path of that module, therefore, it's harder to patch. This way, we get the direct reference to it, and we patch it.

Reply:
Just to clarify, with AWS Chalice it'll typically be an object (class instance), not a class reference. For example with this code snippet:
```
# app.py
from chalice.app import Chalice

app = Chalice(__name__)
```
The Lambda handler module would be `app.app`. That being said though, technically all 3 of the following options are valid callables:
```
# app.py
from typing import Any


class StaticMeta(type):
    @staticmethod
    def __call__(event: dict[str, Any], context: Any) -> dict[str, str]:
        return {"hello": f"from {StaticMeta = }"}


class ClassMeta(type):
    @classmethod
    def __call__(cls, event: dict[str, Any], context: Any) -> dict[str, str]:
        return {"hello": f"from {cls = }"}


class StaticHandler(metaclass=StaticMeta):
    pass


class ClassHandler(metaclass=ClassMeta):
    pass


class InstanceHandler:
    @classmethod
    def __call__(self, event: dict[str, Any], context: Any) -> dict[str, str]:
        return {"hello": f"from {self = }"}


handler1 = StaticHandler
handler2 = ClassHandler
handler3 = InstanceHandler()
```
With the Lambda handler modules being `app.handler1`, `app.StaticHandler`, `app.handler2`, `app.ClassHandler` and `app.handler3`.

Reply:
Thanks @michaeldimchuk, I missed when handler can be called from `@classmethod`, the current state of the PR should work for all of them, but the `ClassHandler`. Will add more testing for this and continue to fully support these use cases.